src/main/java/gov/cabinetofice/gapuserservice/web/LoginControllerV2.java

```diff
@@ -164,22 +92,38 @@ public ModelAndView showNoticePage() {
     }
 
     @GetMapping("/privacy-policy")
-    public ModelAndView showPrivacyPolicyPage(final @ModelAttribute("privacyPolicy") PrivacyPolicyDto privacyPolicyDto) {
+    public ModelAndView submitToPrivacyPolicyPage(final @ModelAttribute("privacyPolicy") PrivacyPolicyDto privacyPolicyDto) {
         return new ModelAndView(PRIVACY_POLICY_PAGE_VIEW)
                 .addObject("homePageUrl", findProperties.getUrl());
     }
 
     @PostMapping("/privacy-policy")
-    public ModelAndView showPrivacyPolicyPage(final @Valid @ModelAttribute("privacyPolicy") PrivacyPolicyDto privacyPolicyDto, final BindingResult result, final HttpServletRequest request, final @CookieValue(name = REDIRECT_URL_COOKIE) Optional<String> redirectUrl) {
-        final Cookie customJWTCookie = WebUtils.getCookie(request, userServiceCookieName);
-        DecodedJWT jwt = JWT.decode(customJWTCookie.getValue());
+    public ModelAndView submitToPrivacyPolicyPage(final @Valid @ModelAttribute("privacyPolicy") PrivacyPolicyDto privacyPolicyDto,
+                                                  final BindingResult result,
+                                                  final HttpServletRequest request,
+                                                  final @CookieValue(name = REDIRECT_URL_NAME) String redirectUrlCookie) {
+        if (result.hasErrors()) return submitToPrivacyPolicyPage(privacyPolicyDto);
+        final User user = getUserFromRequest(request).orElseThrow(() -> new UserNotFoundException("Privacy policy: Could not fetch user from jwt"));
+        return new ModelAndView("redirect:" + runStateMachine(redirectUrlCookie, user));
```

Comment:
Should we use a `RedirectView` instead of `ModelAndView` here?

Reply:
Since line 105 needs to return `ModelAndView` to return validation errors etc, function return has to be that

Reply:
True, missed that :)